src/cache/base.py

```diff
@@ -0,0 +1,102 @@
+from functools import wraps
+from typing import Any, Callable
+from sys import getsizeof
+import pickle
+import time
+
+
+class BaseCache:
+    def __init__(self) -> None:
+        # these specific elements can be change subClasses
+        self.dumper = pickle.dumps
+        self.loader = pickle.loads
+
+    def _log_cache_failed(self, key: str, logger: Any) -> None:
+        """Just some logging"""
+        logger.debug(f">> cache failed on custom_condition for {key=}")
+
+    def _log_cache_hit(self, key: str, logger: Any) -> None:
+        """Just some logging"""
+        logger.debug(f">> cache hit for {key=}")
+
+    def _log_cache_missed(self, key: str, logger: Any) -> None:
+        """Just some logging"""
+        logger.debug(f">> cache missed for {key=}")
+
+    def _size_of(self, value: Any) -> int:
+        """To return the memory size of a value"""
+        return getsizeof(value)
+
+    def _build_key(self, *args: Any, **kwargs: Any) -> str:
+        """The key builder for our cache system"""
+        return f"{args}-{kwargs}"
+
+    def _build_value(self, value: Any) -> bytes:
+        """
+        We need to have an uniform way to build the value and how it will be stored
+        as bytes for perfs requirements
+        """
+        return self.dumper(
+            {
+                "value": value,
+                "at": time.time(),
+                "size_of": self._size_of(value),
+            }
+        )
+
+    def _load_value(self, value: Any) -> Any:
+        return self.loader(value)
+
+    def get(
+        self, key: str, default: Any | None = None, custom_condition: bool = True
+    ) -> Any:
+        """To get something from the cache based on many conditions"""
+        ...
+
+    def set(
+        self, key: str, value: Any | None = None, custom_condition: bool = True
+    ) -> None:
+        """To set something based on condition inside the cache"""
+        ...
```

Comment:
As we are focusing on a data cache, could we think a bit more on what data types we'd like to get in and from such a cache?

Reply:
As you can see on type... i was suggesting that we store dicts(hashmap) in bytes, because it's a key/value datatype, that could lead us storing any kind of datatype we want and fetch it really quickly... or maybe you're talking about something else ?

Reply:
Using dicts in our cache will not be great, because we don't want to manipulate dicts, but rather a more specialized dataframe format in memory.
What counts here is the time to dump and to read data from the disk.

The good point is that we know our data will be in memory as an arrow dataframe at some point, so we better choose a format that suits it. Dict is, in my opinion, way too general.